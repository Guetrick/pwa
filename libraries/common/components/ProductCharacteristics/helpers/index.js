import isMatch from 'lodash/isMatch';

/**
 * Returns the index of a particular characteristic from a set of characteristics.
 * @param {Array} characteristics The characteristics of a product.
 * @param {number} characteristicId The id of the characteristic to find.
 * @return {number}
 */
export function findSelectionIndex(characteristics, characteristicId) {
  const target = characteristics.findIndex(char => char.id === characteristicId);

  if (!target) {
    return -1;
  }

  return target;
}

/**
 * Returns true if the characteristic at the given index is the next characteristic to be selected.
 * @param {Array} characteristics The characteristics of a product.
 * @param {number} index The index of the characteristic to check.
 * @return {boolean}
 */
export function isCharacteristicEnabled(characteristics, index) {
  if (index === 0) {
    return true;
  }

  return !!Object.values(characteristics)[index - 1];
}

/**
 * Returns the selected value by characteristic ID.
 * @param {string} charId The characteristic ID.
 * @param {Object} characteristics The already selected characteristics.
 * @return {string|null}
 */
export function getSelectedValue(charId, characteristics) {
  return characteristics[charId] ? characteristics[charId] : null;
}

/**
 * Prepares the new state after a selection has been made.
 * @param {string} id The selection ID
 * @param {string} value The selection value.
 * @param {Object} selections The selections stored in the state.
 * @param {Array} characteristics The characteristics of a product.
 * @param {Array} products All available products.
 * @return {Object}
 */
export function prepareState(id, value, selections, characteristics, products) {
  const updateValid = !!characteristics.find(({ id: cId, values }) => {
    if (cId !== id) {
      return false;
    }
    return !!values.find(({ id: vId }) => vId === value);
  });

  if (!updateValid) {
    // Input parameters are invalid.
    return selections;
  }

  // Merge the just changed value into the previous selection.
  const currentSelection = {
    ...selections,
    [id]: value,
  };

  if (!selections[id]) {
    // Changed value wasn't set before.
    return currentSelection;
  }

  // Find the current index.
  const index = findSelectionIndex(characteristics, id);

  /**
   * When the given index is the same as the complete set
   * of characteristics then there is nothing to reset.
   */
  if (index === (characteristics.length - 1)) {
    return currentSelection;
  }

  /**
   * Determine if there are products for the current user selection.
   */
  const matches = products.filter(product => (
    isMatch(product.characteristics, currentSelection)
  ));

  if (matches.length === 1) {
    // Product found for the current selection.
    return currentSelection;
  }

  // Find the selections after the given index.
  const after = characteristics.slice(index + 1);

  // Delete the found selections.
  after.forEach((item) => {
    delete currentSelection[item.id];
  });

  return currentSelection;
}
