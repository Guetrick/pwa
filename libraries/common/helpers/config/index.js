import { themeConfig as mock } from './mock';
import pck from './../../package';
/**
 * Provides a default app config as a fallback.
 * @type {Object}
 */
const defaultAppConfig = {
  appId: 'shop_30177',
  browserConnector: false,
  colors: {},
  forgotPasswordUrl: null,
  hasFavorites: false,
  hasReviews: true,
  showWriteReview: true,
  language: 'en-us',
  logo: 'https://example.com/logo',
  marketId: 'US',
  publicPath: 'https://example.com/public',
  shopName: 'Shopgate Connect',
  webCheckoutShopify: null,
  hasNoScanner: false,
  apiUrl: 'https://shop_30177.dev.connect.shopgate.com/',
  shopCNAME: null,
  currency: 'USD',
  showGmdMenuSubHeaders: false,
  benchmark: false,
  sentry: {},
};

export const themeName = process.env.THEME || 'theme';

/**
 * @type {string} package version from package.json
 */
export const pckVersion = pck.version;

/**
 * Provides a default theme config as a fallback.
 * @type {Object}
 */
const defaultThemeConfig = {
  font: {},
  colors: {},
  variables: {},
};

/**
 * Provides a default components config as a fallback.
 * @type {{portals: {}, widgets: {}}}
 */
const defaultComponentsConfig = {
  portals: {},
  widgets: {},
};

/**
 * The app.json config from the theme.
 * @typedef {Object}
 */
const appConfig = process.env.NODE_ENV !== 'test' ? process.env.APP_CONFIG : defaultAppConfig;

/**
 * The components.json config from the theme.
 * @typedef {Object}
 */
export const componentsConfig = {
  ...defaultComponentsConfig,
  ...process.env.COMPONENTS_CONFIG,
};

/**
 * The theme configuration.
 * @typedef {Object}
 */
export const themeConfig = process.env.NODE_ENV === 'test' ? mock : (process.env.THEME_CONFIG || defaultThemeConfig);

/**
 * The shop number.
 * @typedef {string}
 */
const { appId } = appConfig;
export const shopNumber = appId ? appId.replace('shop_', '') : '';

export default appConfig;
