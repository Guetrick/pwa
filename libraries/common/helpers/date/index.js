/**
 * @refactor install moment.js
 */

/**
 * Returns true if date exclusively before (<=).
 * @param {Date} date date object.
 * @param {Date} beforeDate comparison date.
 * @return {boolean}.
 */
export const isBefore = (date, beforeDate) => date < beforeDate;

/**
 * Returns true if date exclusively after (>=).
 * @param {Date} date date object.
 * @param {Date} afterDate comparison date.
 * @return {boolean}.
 */
export const isAfter = (date, afterDate) => date > afterDate;

/**
 * Returns true if date inclusively between (>= & <=).
 * @param {Date} date date object.
 * @param {Date} leftDate left bound.
 * @param {Date} rightDate right bound.
 * @return {boolean}.
 */
export const isBetween = (date, leftDate, rightDate) => (
  date >= leftDate && date <= rightDate
);
