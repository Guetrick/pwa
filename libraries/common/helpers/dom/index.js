/**
 * Checks if the given child is a descendant of the given parent.
 * @param {Object} parent The parent element.
 * @param {Object} child The child element.
 * @return {boolean}
 */
export const isDescendant = (parent, child) => {
  let node = child.parentNode;

  while (node !== null) {
    if (node === parent) return true;
    node = node.parentNode;
  }

  return false;
};

/**
 * Retrieves the offset of a element.
 * @param {Object} element The element to retrieve the offset for.
 * @returns {Object} The offset of the element contain `top` and `left` position values.
 */
export const getOffset = element => element.getBoundingClientRect();

/**
 * Calculate the height of an element including it's y-axis margins.
 * @param {HTMLElement} element The DOM element.
 * @returns {number} The absolute height of the element.
 */
export const getAbsoluteHeight = (element) => {
  // Get the styles of the element.
  const styles = window.getComputedStyle(element);
  // Isolate the y-axis margins.
  const margins = parseFloat(styles.marginTop) + parseFloat(styles.marginBottom);
  // Add the margins to the element's height and return it.
  return Math.ceil(element.offsetHeight + margins);
};
