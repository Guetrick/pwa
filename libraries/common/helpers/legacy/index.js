import hideMenuBar from '@shopgate/pwa-core/commands/hideMenuBar';
import broadcastEvent from '@shopgate/pwa-core/commands/broadcastEvent';
import showTab from '@shopgate/pwa-core/commands/showTab';

/**
 * Page context that comes from the app.
 * @type {Object}
 */
let currentPageContext = {};

/**
 * Get Page context if available.
 * @returns {Object}
 */
export const getPageContext = () => currentPageContext;

/**
 * Prepares the NavigationBar and the MenuBar of the app legacy part for the PWA.
 */
export const prepareLegacyNavigation = () => {
  /**
   * Broadcasts an event to the pwa_navigation_bar webview and updates the navigation bar with
   * type "none". Event parameters are defined accordingly to the specification of the native
   * updateNavigationBar event.
   */
  broadcastEvent({
    event: 'updateNavigationBar',
    parameters: ['', 'none', { type: 'none' }],
  });

  hideMenuBar({ animation: null });
};

/**
 * Shows the main tab when legacy system requests the previous tab.
 * @param {Object} context The page context from the app.
 * @param {string} context.tab The tab where the page is opened.
 */
export const pageContext = (context) => {
  currentPageContext = context;
};

/**
 * Shows the main tab when legacy system requests the previous tab.
 */
export const showPreviousTab = () => {
  showTab({ targetTab: currentPageContext.tab ? currentPageContext.tab : 'main' });
};
