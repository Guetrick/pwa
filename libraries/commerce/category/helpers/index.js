import { bin2hex } from '@shopgate/pwa-common/helpers/data';
import { CATEGORY_PATH } from '../constants';

/**
 * Generate category route for navigation.
 * @param {string} id category Id
 * @returns {string}
 */
export const getCategoryRoute = id => `${CATEGORY_PATH}/${bin2hex(id)}`;

