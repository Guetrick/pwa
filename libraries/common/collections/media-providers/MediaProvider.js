import { logger } from '@shopgate/pwa-core/helpers';
import styles from './style';

/**
 * The MediaProvider base class.
 */
class MediaProvider {
  /**
   * Constructor.
   */
  constructor() {
    this.containers = new Map();
  }

  /**
   * Optimizes video container to make it responsive.
   * @param {NodeList} container A DOM container.
   * @private
   * @returns {MediaProvider}
   */
  responsify(container) {
    // Remove fixed dimensions from the container.
    container.removeAttribute('height');
    container.removeAttribute('width');

    // Create the wrapper and apply styling.
    const wrapper = document.createElement('div');

    wrapper.className = styles;

    // Add the wrapper right before the container into the DOM.
    container.parentNode.insertBefore(wrapper, container);
    // Move the container into the wrapper.
    wrapper.appendChild(container);

    return this;
  }

  /**
   * Add a DOM container with embedded videos.
   * @param {NodeList} container A DOM container.
   * @returns {MediaProvider}
   */
  add() {
    logger.error('MediaProvider.add() needs to be implemented within an inheriting class');
    return this;
  }

  /**
   * Remove a DOM container.
   * @param {NodeList} container A DOM container.
   * @returns {MediaProvider}
   */
  remove(container) {
    this.containers.delete(container);
    return this;
  }

  /**
   * Stops all playing videos within the DOM containers.
   * @returns {MediaProvider}
   */
  stop() {
    logger.error('MediaProvider.stop() needs to be implemented within an inheriting class');
    return this;
  }
}

export default MediaProvider;
