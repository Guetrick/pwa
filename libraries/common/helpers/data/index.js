import { logger } from '@shopgate/pwa-core/helpers';
import { isObject, isArray } from '../validation';

/**
 * Tests if the prop is an object or an array.
 * @param {*} prop The property to test.
 * @return {boolean}
 */
const isObjectOrArray = prop => isObject(prop) || isArray(prop);

/**
 * Remove the first element of the array.
 * @param {Array} array The array to remove the first element from.
 * @returns {Array} The reduced array.
 */
export const shift = ([, ...newArray]) => newArray;

/**
 * Returns a new object without certain keys.
 * @param {Object} obj The original object.
 * @param {Array} [keys=[]] An array of unwanted keys.
 * @return {Object} The reduced object.
 */
export const objectWithoutProps = (obj, keys = []) => {
  const target = {};

  Object.keys(obj).forEach((key) => {
    if (!keys.includes(key)) {
      target[key] = obj[key];
    }
  });

  return target;
};

/**
 * Checks if a URL is an external URL.
 * @param {string} url The URL to check.
 * @return {boolean}
 */
export const isExternal = url =>
  url.includes('http://') || url.includes('https://') || url.includes('//');

/**
 * Returns the actual url to the image, by adding url parameters with the dimensions for img-cdn
 * @param {string} src Source to the image.
 * @param {Object} dimension Dimension of the requested image.
 * @param {number} dimension.width Width in pixels.
 * @param {number} dimension.height Height in pixels.
 * @returns {string}
 */
export const getActualImageSource = (src, { width, height }) => {
  if (src && src.startsWith('https://img-cdn.shopgate.com') && !src.includes('?')) {
    return `${src}?w=${width}&h=${height}&q=70&zc=resize&fillc=FFFFFF`;
  }
  return src;
};

/**
 * Finds the index of a property inside an array.
 * @param {Array} arr The array to check.
 * @param {Function} fn The callback function to filter inside the array.
 * @return {number} The index of the property.
 */
export const findIndex = (arr, fn) => {
  if (Array.isArray(arr)) {
    for (let i = 0; i < arr.length; i += 1) {
      if (fn(arr[i])) {
        return i;
      }
    }
  }

  return -1;
};

/**
 * Convert binary data into hexadecimal representation
 * @param {string|number} str The string that shall be encoded
 * @see http://locutus.io/php/strings/bin2hex/
 * @return {string} The hexadecimal representation of the given string
 */
export const bin2hex = (str) => {
  const s = `${str}`;

  let i;
  let l;
  let o = '';
  let n;

  for (i = 0, l = s.length; i < l; i += 1) {
    n = s.charCodeAt(i).toString(16);
    o += n.length < 2 ? `0${n}` : n;
  }

  return o;
};

/**
 * Decodes a hexadecimal encoded binary string
 * @param {string} str The string that shall be decoded
 * @see http://locutus.io/php/strings/hex2bin/
 * @returns {string|boolean} Hexadecimal representation of data. FALSE if decoding failed.
 */
export const hex2bin = (str) => {
  const s = `${str}`;

  const ret = [];
  let i = 0;
  let l;

  for (l = s.length; i < l; i += 2) {
    const c = parseInt(s.substr(i, 1), 16);
    const k = parseInt(s.substr(i + 1, 1), 16);

    if (Number.isNaN(c) || Number.isNaN(k)) {
      return false;
    }

    // eslint-disable-next-line no-bitwise
    ret.push((c << 4) | k);
  }

  // eslint-disable-next-line prefer-spread
  return String.fromCharCode.apply(String, ret);
};

/**
 * Compares two strings of object keys for object sorting.
 * @param {string} a The first key.
 * @param {string} b The second key.
 * @return {number} The sort order.
 */
const defaultKeySortFn = (a, b) => a.localeCompare(b);

/**
 * Deep sorts an object by its keys.
 * @param {Object} src The object to sort.
 * @param {Function} [comparator=defaultKeySortFn] The sorting operator callback.
 * @return {Object} The sorted object.
 */
export const sortObject = (src, comparator = defaultKeySortFn) => {
  if (Array.isArray(src)) {
    return src.map(item => sortObject(item, comparator));
  }

  if (isObject(src)) {
    return Object.keys(src)
      .sort(comparator)
      .reduce(
        (obj, key) => ({
          ...obj,
          [key]: sortObject(src[key], comparator),
        }),
        {}
      );
  }

  return src;
};

/**
 * Transforms widget display options to match pipeline specifications.
 * @param {string} displayOptions The display options as specified by the widget settings.
 * @returns {string} The transformed string compatible with the pipeline requests.
 */
export const transformDisplayOptions = (displayOptions) => {
  switch (displayOptions) {
    case 'relevance_asc':
    case 'relevance_desc':
      return 'relevance';
    case 'price_asc':
      return 'priceAsc';
    case 'price_desc':
      return 'priceDesc';
    case 'name_asc':
      return 'nameAsc';
    case 'name_desc':
      return 'nameDesc';
    default:
      return displayOptions;
  }
};

/**
 * Validates all parameters. The selector will return null if one parameter is invalid.
 * @param {Function} selector The original selector.
 * @param {Object} [defaultResult] The result when the selector fails.
 * @return {Function}
 */
export const validateSelectorParams = (selector, defaultResult = null) => (...params) => {
  if (params.some(param => param === null || typeof param === 'undefined')) {
    return defaultResult;
  }

  return selector(...params);
};

/**
 * Takes a destination object and a source and merges the source object into the destination.
 * Differing properties will
 * @param {Object} destination Object to mutate
 * @param {Object} source Object which contains the properties to assign to the destination
 * @param {boolean} [warn] Enables log output on mismatching types. Defaults to 'true'.
 */
export function assignObjectDeep(destination, source, warn = true) {
  // Avoids eslint warning on param mutation, which is necessary
  const dest = destination;
  const src = source;

  // Reassign is needed here because the destination param must be mutated insteatd of.
  if (!isObject(dest) || Array.isArray(dest) || !isObject(src) || Array.isArray(src)) {
    logger.error('Operands for "assignObjectDeep" must be objects');
    return;
  }

  Object.keys(src).forEach((key) => {
    const prop = src[key];
    if (!isObjectOrArray(dest[key]) || !isObjectOrArray(prop)) {
      // output a warning if only one of both is an object (undefined dest is fine -> no warning)
      if (warn && dest[key] !== undefined
        && (!isObjectOrArray(dest[key]) ? isObjectOrArray(prop) : isObjectOrArray(prop))) {
        logger.warn('Trying to merge object properties with mixed object types: ', prop, dest[key]);
      }

      // Overwrite always if one of the props is not an object or array
      dest[key] = prop;
      return;
    }

    // Both structures are objects but one or both can be an array
    if (isArray(prop) && isArray(dest[key])) {
      // Merge arrays
      prop.forEach((element) => { dest[key].push(element); });
    } else if (!isArray(prop) && !isArray(dest[key])) {
      // Merge objects
      assignObjectDeep(dest[key], prop);
    } else {
      // Object types differ, print a warning
      if (warn) {
        logger.warn('Merging object properties with mixed object types: ', prop, dest[key]);
      }

      dest[key] = prop;
    }
  });
}
