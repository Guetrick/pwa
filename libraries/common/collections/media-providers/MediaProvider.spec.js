import { JSDOM } from 'jsdom';
import { logger } from '@shopgate/pwa-core/helpers';
import MediaProvider from './MediaProvider';

jest.mock('@shopgate/pwa-core/helpers', () => ({
  logger: {
    error: jest.fn(),
  },
}));

/**
 * Creates a DOM container with iframes.
 * @param {Array} srcs A list of video URLs.
 * @return {Object}
 */
const createContainer = (srcs) => {
  const html = srcs.map(src => `<iframe src="${src}"></iframe>`).join('');
  return new JSDOM(html).window.document;
};

describe('MediaProvider', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('.constructor()', () => {
    it('should construct as expected', () => {
      const instance = new MediaProvider();
      expect(instance.containers).toBeInstanceOf(Map);
      expect(Array.from(instance.containers)).toHaveLength(0);
    });
  });

  describe('.add()', () => {
    it('should log an error when it is not overwritten', () => {
      const instance = new MediaProvider();
      instance.add();
      expect(logger.error).toHaveBeenCalledTimes(1);
    });
  });

  describe('.stop()', () => {
    it('should log an error when it is not overwritten', () => {
      const instance = new MediaProvider();
      instance.stop();
      expect(logger.error).toHaveBeenCalledTimes(1);
    });
  });

  describe('.remove()', () => {
    it('should remove containers as expected', () => {
      const instance = new MediaProvider();
      const containerOne = createContainer(['http://www.provider-one.com/video']);
      const containerTwo = createContainer(['http://www.provider-two.com/video']);

      instance.containers.set(containerOne, containerOne);
      instance.containers.set(containerTwo, containerTwo);

      instance.remove(containerTwo);

      expect(instance.containers.size).toBe(1);
      expect(instance.containers.get(containerTwo)).toBeUndefined();
      expect(instance.containers.get(containerOne)).toEqual(containerOne);

      instance.remove(containerOne);

      expect(instance.containers.size).toBe(0);
    });
  });

  describe.skip('.responsify()', () => {
    it('should optimize a container to be responsive', () => {
      // TODO: Implement the test when a solution for the insertBefore issue was found.
    });
  });
});
