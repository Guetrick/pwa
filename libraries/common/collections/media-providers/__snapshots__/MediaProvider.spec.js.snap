// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaProvider .responsify() should responsify a dom container 1`] = `
<body>
  <div
    class="css-2gg043"
  >
    <iframe />
  </div>
</body>
`;
